File: backend/app/api/v1/docs.py
"""Document endpoints on ``/api/v1/docs`` (M4; T14; ADR-0019).

Routes (all read-only in M4):

    GET  /docs            viewer+  list generated documents (paginated, filterable)
    GET  /docs/{doc_id}   viewer+  detail (includes content)
    GET  /docs/{doc_id}/download  viewer+  raw content download payload

Documents are agent-generated artifacts (inventory / diagram / runbook).  The
A9 redaction layer is applied at the LLM boundary before a document is written
(ADR-0019), so the stored ``content`` is already redacted and safe to return to
any authenticated viewer.

The router tag is ``"documents"``; the prefix ``/docs`` is set here so that
``app.api.v1.__init__`` can include it without repeating the prefix.
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, Query
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db, require_role
from app.core.errors import NotFoundError
from app.models import Document, DocumentKind, User
from app.schemas.config_mgmt import (
    DocumentDownload,
    DocumentListResponse,
    DocumentRead,
)

router = APIRouter(prefix="/docs", tags=["documents"])

DbSession = Annotated[AsyncSession, Depends(get_db)]
Viewer = Annotated[User, Depends(require_role("viewer"))]


@router.get("", response_model=DocumentListResponse)
async def list_documents(
    session: DbSession,
    _user: Viewer,
    kind: Annotated[DocumentKind | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=500)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> DocumentListResponse:
    """List generated documents, optionally filtered by kind (inventory/diagram/runbook).

    Viewer+ may access this endpoint.  Documents are agent-generated and the
    A9 redaction layer was applied at write time (ADR-0019), so content is safe
    for any authenticated user.
    """
    base = select(Document)
    if kind is not None:
        base = base.where(Document.kind == kind)

    total = (await session.execute(select(func.count()).select_from(base.subquery()))).scalar_one()
    rows = (
        (
            await session.execute(
                base.order_by(Document.generated_at.desc(), Document.id).limit(limit).offset(offset)
            )
        )
        .scalars()
        .all()
    )
    return DocumentListResponse(
        items=[DocumentRead.model_validate(r) for r in rows],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/{doc_id}", response_model=DocumentRead)
async def get_document(
    doc_id: uuid.UUID,
    session: DbSession,
    _user: Viewer,
) -> DocumentRead:
    """Return one generated document by id (404 when unknown)."""
    doc = await session.get(Document, doc_id)
    if doc is None:
        raise NotFoundError(f"document {doc_id} does not exist")
    return DocumentRead.model_validate(doc)


@router.get("/{doc_id}/download", response_model=DocumentDownload)
async def download_document(
    doc_id: uuid.UUID,
    session: DbSession,
    _user: Viewer,
) -> DocumentDownload:
    """Return the download payload for a generated document (title + format + content).

    Viewer+ may access this endpoint.  This is the canonical download path for
    the frontend document library (M4 T16).
    """
    doc = await session.get(Document, doc_id)
    if doc is None:
        raise NotFoundError(f"document {doc_id} does not exist")
    return DocumentDownload.model_validate(doc)

File: frontend/src/api/docs.ts
/**
 * Typed client functions for the M4 document-library endpoints (T14 / T16).
 *
 * Routes (all read-only; viewer+ RBAC; ADR-0019):
 *   GET /docs                  viewer+  list generated documents (paginated, filterable)
 *   GET /docs/{id}             viewer+  document detail (includes content)
 *   GET /docs/{id}/download    viewer+  download payload (title + format + content)
 *
 * Mirrors ``backend/app/schemas/config_mgmt.py`` DocumentRead / DocumentDownload
 * and ``backend/app/api/v1/docs.py``.
 */

import { apiFetch } from "./client";

// ── Enum literals (match backend DocumentKind / DocumentFormat) ───────────────

export type DocumentKind = "inventory" | "diagram" | "runbook" | "incident_report";
export type DocumentFormat = "md" | "csv" | "mermaid";

// ── Response shapes ───────────────────────────────────────────────────────────

export interface DocumentRead {
  id: string;
  kind: DocumentKind;
  title: string;
  format: DocumentFormat;
  content: string;
  source_refs: Record<string, unknown>;
  generated_at: string;
  generated_by_session_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentListResponse {
  items: DocumentRead[];
  total: number;
  limit: number;
  offset: number;
}

export interface DocumentDownload {
  id: string;
  title: string;
  format: DocumentFormat;
  content: string;
  generated_at: string;
}

// ── Query-string params ───────────────────────────────────────────────────────

export interface ListDocumentsParams {
  kind?: DocumentKind;
  limit?: number;
  offset?: number;
}

// ── API functions ─────────────────────────────────────────────────────────────

/** ``GET /api/v1/docs`` — paginated list, optionally filtered by kind. */
export function listDocuments(
  params: ListDocumentsParams = {},
): Promise<DocumentListResponse> {
  const qs = new URLSearchParams();
  if (params.kind !== undefined) qs.set("kind", params.kind);
  if (params.limit !== undefined) qs.set("limit", String(params.limit));
  if (params.offset !== undefined) qs.set("offset", String(params.offset));
  const query = qs.toString();
  return apiFetch<DocumentListResponse>(`/docs${query ? `?${query}` : ""}`);
}

/** ``GET /api/v1/docs/{id}`` — document detail including content. */
export function getDocument(docId: string): Promise<DocumentRead> {
  return apiFetch<DocumentRead>(`/docs/${docId}`);
}

/** ``GET /api/v1/docs/{id}/download`` — download payload (title + format + content). */
export function downloadDocument(docId: string): Promise<DocumentDownload> {
  return apiFetch<DocumentDownload>(`/docs/${docId}/download`);
}
